src/pages/graphile-build/hooks.md

```diff
@@ -20,54 +21,42 @@ const MyType = newWithHooks(GraphQLObjectType, spec);
 
-The `hookName` that you register via `builder.hook(hookName, hookFunction)`
-must match one of the supported hooks. See [All
-Hooks](/graphile-build/all-hooks/) for a list of all the hooks we support,
-here's a brief overview of some of the more important ones:
+### Stages of the build process
 
-- `build`: extend the [Build object](/graphile-build/build-object/) passed to all other hooks
+The `hookName` that you register the function for must match [one of the supported hooks](/graphile-build/all-hooks/).
 
-- `init`: perform setup after `build` freezes but before building the schema starts
+Here's a brief rundown of how the hooks are used by the Graphile engine.
+See the [`buildSchema`](https://github.com/graphile/graphile-engine/blob/v4.4.4/packages/graphile-build/src/SchemaBuilder.js#L474-L499)
+and [`createBuild`](https://github.com/graphile/graphile-engine/blob/v4.4.4/packages/graphile-build/src/SchemaBuilder.js#L450-L472) methods in the source code.
 
-- `GraphQLSchema`: root-level schema - hook to add `query`,
-  `mutation` or `subscription` fields; called implicitly by `buildSchema(plugins, options)`
+1.  A new [Build object](/graphile-build/build-object/) with the basic functionality is created
+    by [`makeNewBuild`](https://github.com/graphile/graphile-engine/blob/v4.4.4/packages/graphile-build/src/makeNewBuild.js#L230)
+2.  The `inflection` hook allows plugins to add new inflection methods to the `build.inflection` object, or overwrite previously declared ones.
+3.  The `build` hook allows plugins to add new utility methods to the `build` object itself, or overwrite previously declared ones.
+4.  The `init` hook is the setup phase where the schema types get created by the plugins,
+    using the [`newWithHooks`](/graphile-build/build-object/#newwithhookstype-spec-scope)
+    or [`addType`](/graphile-build/build-object/#addtypetype-graphqlnamedtype) methods of the build object.
+    For example, the [default `StandardTypesPlugin`](/graphile-build/default-plugins/#standardtypesplugin) registers the builtin scalars,
+    or the PostGraphile plugins register the types found in the introspection results.
+5.  The schema is constructed using `newWithHooks(GraphQLSchema, …)`, where the `query`, `mutation` and `subscription` root operations
+    are configured by the respective [default plugins](/graphile-build/default-plugins)
+    and [other schema options](https://github.com/graphql/graphql-js/blob/v14.5.6/src/type/schema.js#L318-L324) can be adjusted.
+6.  The `finalize` hook allows plugins to alter the schema instance and all its contents.
 
-- When creating a `GraphQLObjectType` via
-  `newWithHooks`:
+During the `init` hook, the [hooks for the respective GraphQL types](/graphile-build/all-hooks/) are applied [in the
+`newWithHooks` function](https://github.com/graphile/graphile-engine/blob/v4.4.4/packages/graphile-build/src/makeNewBuild.js#L329),
```

Comment:
I think footnotes might be more appropriate for this level of detail - I don't think people starting with the library necessarily want links to the source.

Reply:
I found it very useful to understand how the hook system works - of course ideally documentation will be good enough so that people don't have to read the source any more. Should I rather link to https://www.graphile.org/graphile-build/build-object/#newwithhookstype-spec-scope and put the details there?

Reply:
It's a philosophical thing; I think the docs should only call out to the underlying source code when that's critically important in what they're doing - for example writing inflectors. The rest of the time, the documentation should be standalone.

Adding embellishments about how it works internally is totally fine _in a footnote_ but it shouldn't be in the main body of the text. People shouldn't feel that they have to read the source in order to know how to use it. Also the source should evolve over time, but the docs should be stable - the docs outline the intended behaviour and interface, whereas the code outlines the _actual_ behaviour/interface, which is a moving target.

Reply:
Is there a syntax for footnotes in the used markdown engine?

Reply:
Apparently this syntax should work; but I've not tried it:

```md
This is normal body copy.[^also] It includes a couple footnotes.[^thing]

[^also]:
  This is a footnote.

[^thing]:
  This is another footnote.
```

Reply:
We're using gatsby; so you can search "gatsby footnote" and maybe experiment with some of those plugins if you're so inclined :+1: